plugin/pkg/webhook/webhook.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package webhook implements the authorizer.Authorizer interface using HTTP webhooks.
+package webhook
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer/json"
+	"k8s.io/kubernetes/pkg/runtime/serializer/versioning"
+
+	_ "k8s.io/kubernetes/pkg/apis/authorization/install"
+)
+
+type GenericWebhook struct {
+	RestClient *restclient.RESTClient
+}
+
+// New creates a new GenericWebhook from the provided kubeconfig file.
+func NewGenericWebhook(kubeConfigFile string, groupVersions []unversioned.GroupVersion) (*GenericWebhook, error) {
+	for _, groupVersion := range groupVersions {
+		if !registered.IsEnabledVersion(groupVersion) {
+			return nil, fmt.Errorf("webhook plugin requires enabling extension resource: %s", groupVersion)
+		}
+	}
+
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.ExplicitPath = kubeConfigFile
+	loader := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+
+	clientConfig, err := loader.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	serializer := json.NewSerializer(json.DefaultMetaFactory, api.Scheme, runtime.ObjectTyperToTyper(api.Scheme), false)
```

Comment:
Yeah, anyone doing json.NewSerializer like the should be using LegacyCodec
- very few people should ever be using anything except a)
NegotiatedSerializer or b) api.Codecs.LegacyCodec()

On Thu, Apr 28, 2016 at 5:58 PM, CJ Cullen <notifications@github.com> wrote:

> In plugin/pkg/webhook/webhook.go
> <https://github.com/kubernetes/kubernetes/pull/24902#discussion_r61509591>
> :
>
> > +func NewGenericWebhook(kubeConfigFile string, groupVersions []unversioned.GroupVersion) (*GenericWebhook, error) {
> > +	for _, groupVersion := range groupVersions {
> > +		if !registered.IsEnabledVersion(groupVersion) {
> > +			return nil, fmt.Errorf("webhook plugin requires enabling extension resource: %s", groupVersion)
> > +		}
> > +	}
> > +
> > +	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
> > +	loadingRules.ExplicitPath = kubeConfigFile
> > +	loader := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
> > +
> > +	clientConfig, err := loader.ClientConfig()
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	serializer := json.NewSerializer(json.DefaultMetaFactory, api.Scheme, runtime.ObjectTyperToTyper(api.Scheme), false)
>
> I just pulled this over from the webhook authorizer. I can change it to
> use the LegacyCode if that is a better approach.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24902/files/765d75020629d41818092cba9da5d6a1c4ee3287#r61509591>
>